text/0012-corner-smoothing.md

```diff
@@ -0,0 +1,266 @@
+Windows, macOS, and Linux.
+-->
+
+The corners of elements rounded by the CSS `border-radius` rule can be smoothed out using the `-electron-corner-smoothing` feature.
+
+<table>
+<tr>
+<th>Code</th>
+<th>Result</th>
+</tr>
+<tr>
+<td>
+<pre lang="css">
+.box {
+  width: 128px;
+  height: 128px;
+  border-radius: 24px;
+  -electron-corner-smoothing: 100%;
+}
+</pre>
+</td>
+<td>
+<img src="../images/0012/Rectangle.svg" width="128" alt="" />
+</td>
+</tr>
+</table>
+
+The `-electron-corner-smoothing` CSS rule is **only implemented for Electron** and has no effect in browsers. Avoid using this rule outside of Electron.
+
+The `-electron-corner-smoothing` CSS rule controls how far the corner's curvature can extend into the edge of the element. It is a percentage between `0%` (default) and `100%`. `-electron-corner-smoothing` affects the shape of borders, outlines, and shadows on the target element.
+
+To match the system rounding in macOS, use a value of `60%`. `-electron-corner-smoothing` is available across all Electron platforms.
+
+![A chart demonstrating values of the CSS rule and their visual effects on the same rectangle. The values of 0%, 30%, 60%, and 100% are demonstrated, with the 60% value being labeled as macOS.](../images/0012/Values.svg)
+
+This smooth rounding is similar to Apple's "continuous" rounded corners in SwiftUI as well as Figma's "corner smoothing" control on design elements.
+
+## Reference-level explanation
+
+Integration with macOS's design language is a primary goal for this proposal. Enabling design customization is a secondary goal.
+
+To implement corner smoothing we will patch Blink with two additions: a CSS property and support for painting smooth corners.
+
+### CSS Property
+
+CSS properties are [declared in a large JSON list](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/css/css_properties.json5). We will create a new CSS rule by adding a new entry to this list and configuring its properties appropriately. Doing so automatically generates C++ classes and glue code at compile time.
+
+The `-electron-` vendor prefix was added to the CSS rule to [satsify some concerns](https://www.w3.org/TR/css-2023/#proprietary). It clearly identifies this rule as an Electron-specific extension to developers, removing the expectation that other platforms will also implement this extension.
+
+### Painting
+
+The painting procedures in Blink are defined across multiple "Painter" classes that use the [`GraphicsContext` class](https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/platform/graphics/graphics_context.h) to render graphics. The API is very similar to the Web's [2D canvas rendering API](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D).
+
+We will add new logic to these Painter classes to identify elements with smooth corners, then use a new procedure to paint them.
```

Comment:
It might be worth taking a look at [skia_renderer.cc](https://source.chromium.org/chromium/chromium/src/+/main:components/viz/service/display/skia_renderer.cc;l=1094;drc=68187a1957ea6d5a3169617a9b141a45bb210dc5) as well. Last I checked, all supported Chromium platforms have moved to this renderer (from their DirectX/OpenGL renderers). There are references to rounded corners here as a low-level optimization for masking. Might have some effect in creating graphical artifacts if the rounding masks differ.